nexus/db-queries/src/db/datastore/deployment.rs

```diff
@@ -0,0 +1,1578 @@
+///         --
+///         -- If either of these cases fails, we return `parent-not-target`.
+///         (
+///            SELECT "parent_blueprint_id" FROM "blueprint", current_target
+///            WHERE
+///              "id" = <new_target_id>
+///              AND current_target.blueprint_id = "parent_blueprint_id"
+///         ) IS NOT NULL
+///         OR
+///         (
+///            SELECT 1 FROM "blueprint"
+///            WHERE
+///              "id" = <new_target_id>
+///              AND "parent_blueprint_id" IS NULL
+///              AND NOT EXISTS (SELECT version FROM current_target)
+///         ) = 1,
+///         <new_target_id>,
+///         'parent-not-target'
+///       )
+///     ) AS UUID)
+///   ),
+///
+///   -- Determine the new version number to use: either 1 if this is the
+///   -- first blueprint being made the current target, or 1 higher than
+///   -- the previous target's version.
+///   --
+///   -- The final clauses of each of these WHERE clauses repeat the
+///   -- checks performed above in `check_validity`, and will cause this
+///   -- subquery to return no rows if we should not allow the new
+///   -- target to be set.
+///   new_target AS (
+///     SELECT 1 AS new_version FROM "blueprint"
+///       WHERE
+///         "id" = <new_target_id>
+///         AND "parent_blueprint_id" IS NULL
+///         AND NOT EXISTS (SELECT version FROM current_target)
+///     UNION
+///     SELECT current_target.version + 1 FROM current_target, "blueprint"
+///       WHERE
+///         "id" = <new_target_id>
+///         AND "parent_blueprint_id" IS NOT NULL
+///         AND "parent_blueprint_id" = current_target.blueprint_id
+///   )
+///
+///   -- Perform the actual insertion.
+///   INSERT INTO "bp_target"(
+///     "version","blueprint_id","enabled","time_made_target"
+///   )
+///   SELECT
+///     new_target.new_version,
+///     <new_target_id>,
+///     <new_target_enabled>,
+///     <new_target_time_made_target>
+///     FROM new_target
+/// ```
+#[derive(Debug, Clone, Copy)]
+struct InsertTargetQuery {
+    target_id: Uuid,
+    enabled: bool,
+    time_made_target: DateTime<Utc>,
+}
+
+// Uncastable sentinel used to detect we attempt to make a blueprint the target
+// when it does not exist in the blueprint table.
+const NO_SUCH_BLUEPRINT_SENTINEL: &str = "no-such-blueprint";
+
+// Uncastable sentinel used to detect we attempt to make a blueprint the target
+// when its parent_blueprint_id is not the current target.
+const PARENT_NOT_TARGET_SENTINEL: &str = "parent-not-target";
+
+// Error messages generated from the above sentinel values.
+const NO_SUCH_BLUEPRINT_ERROR_MESSAGE: &str =
+    "could not parse \"no-such-blueprint\" as type uuid: \
+     uuid: incorrect UUID length: no-such-blueprint";
+const PARENT_NOT_TARGET_ERROR_MESSAGE: &str =
+    "could not parse \"parent-not-target\" as type uuid: \
+     uuid: incorrect UUID length: parent-not-target";
```

Comment:
I am not sure these are stable messages.  But I guess as long as we test the behavior, we should be fine.

Reply:
Agreed on both counts. FWIW this is not a new technique; I cribbed it from https://github.com/oxidecomputer/omicron/blob/d5dace65fa093174a7502e0f6a3dde4ccabe6337/nexus/db-queries/src/db/queries/network_interface.rs#L89-L94, which also has tests covering it.